pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/GrpcMailboxService.java

```diff
@@ -44,16 +48,17 @@
  * </ul>
  */
 public class GrpcMailboxService implements MailboxService<TransferableBlock> {
+  private static final Duration MAILBOX_CACHE_EXPIRY = Duration.ofMinutes(5);
   // channel manager
   private final ChannelManager _channelManager;
   private final String _hostname;
   private final int _mailboxPort;
 
   // maintaining a list of registered mailboxes.
-  private final ConcurrentHashMap<String, ReceivingMailbox<TransferableBlock>> _receivingMailboxMap =
-      new ConcurrentHashMap<>();
-  private final ConcurrentHashMap<String, SendingMailbox<TransferableBlock>> _sendingMailboxMap =
-      new ConcurrentHashMap<>();
+  private final Cache<String, ReceivingMailbox<TransferableBlock>> _receivingMailboxCache =
+      CacheBuilder.newBuilder().expireAfterAccess(MAILBOX_CACHE_EXPIRY.toMinutes(), TimeUnit.MINUTES).build();
```

Comment:
This doesn't work based on the doc here:

If [expireAfterWrite](https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html#expireAfterWrite(long,%20java.util.concurrent.TimeUnit)) or [expireAfterAccess](https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html#expireAfterAccess(long,%20java.util.concurrent.TimeUnit)) is requested entries may be evicted on each cache modification, on occasional cache accesses, or on calls to [Cache.cleanUp()](https://guava.dev/releases/18.0/api/docs/com/google/common/cache/Cache.html#cleanUp()). Expired entries may be counted by [Cache.size()](https://guava.dev/releases/18.0/api/docs/com/google/common/cache/Cache.html#size()), but will never be visible to read or write operations.

Also the cleanup thread needs to be tied with query deadline.

Reply:
@ankitsultana ^

Reply:
@61yao how does it "not work"? The periodic cache maintenance tasks are run on writes or occasionally during reads. Only if the cache is not accessed for a few hours would the entry not be cleaned in time. Instead of a Guava cache we could consider using a different strategy but since this is only temporary I weighed on the side of introducing as little temporary code as possible.

> Also the cleanup thread needs to be tied with query deadline.

I have set the expiry to 5 minutes which should be larger than any query deadline.

Reply:
I see what you meant. the read and write access will clean up expired entries. 

"larger than any query deadline" is your assumption. 

What's the downside of passing the actual deadline?